topics/dev/tutorials/debugging/tutorial.md

```diff
@@ -0,0 +1,730 @@
+>    >  
+>    >  The solution is hidden below, but try your hand at fixing it first:
+>    >  
+>    >  > ### {% icon solution %} Solution
+>    >  >
+>    >  > The failure is only happening on the disabled test case, so take a look in the template of `SavedRuleSelector` where the conditions for adding disabled are defined.  
+>    >  > The way it's worded in the test (and the way one could think about it logically), is that if there is no history, the button should be disabled, but the code says 
+>    >  > ```javascript
+>    >  > :class="{ disabled: numOfSavedRules == 1 }"
+>    >  > ``` 
+>    >  > -- whoops! A very silly off by one error. On that very same line, change the 1 to a 0, and run the test again. 
+>    >  {: .solution }
+>    >  
+>    >  If you think you've got it, try running the test locally again to be sure.
+>    >  
+>    >  > ### {% icon comment %} Running fewer client tests
+>    >  >
+>    >  > If you don't want to run the whole suite of client tests, you can add keywords that match the test path/name. For example both `yarn run jest rule` and `yarn jest selector` will work, but the former will run other tests with rule in the path/module name.  
+>    >  {: .comment}
+>    >  
+> 4. __Finishing up__
+>    >  
+>    >  Nice Job! If you're here, you've found the problem, ran it locally, fixed the problem, ran it locally again without errors, and are ready to commit your change. Write a useful commit message like `Fix logic error in SavedRuleSelector` or something similar, making sure it is clear what you've done.
+>    > (See [Galaxy's contributor's guide](https://github.com/galaxyproject/galaxy/blob/dev/CONTRIBUTING.md#how-to-contribute) for details on how to write a useful and properly formatted commit message.)
+>    >  
+>    >  Awesome work - you now understand how to solve client linting issues!- you now have a handle on client tests!
+{: .hands_on}
+
+# Selenium test failure
+
+Selenium is the end-to-end or integration testing framework that we use; so these tests often are designed to test both frontend and backend code. This makes fixing them a little difficult, but we are sure you can do it!
+
+> ### {% icon hands_on %} Hands-on: Fixing a failing Selenium test
+> 
+> 1. __Finding the failing test on GitHub__
+>    >  
+>    >  There may be actually a few failing Selenium tests, but the good new is, they are the result of only one little error. The failing tests on GitHub says "Selenium Test 3.7, 2". When you click on the details of this one, it's not always very clear which test is causing the failure, but if you download the artifacts, it's very clear which tests have failed.
+>    >  
+>    >  Try to identify which tests are failing based on artifacts you downloaded.
+>    >  
+>    >  > ### {% icon solution %} Solution
+>    >  >
+>    >  > The failing tests are:
+>    >  >
+>    >  > - `test_build_list_and_show_items`
+>    >  > - `test_data_column_input`
+>    >  > - `test_upload_pair_specify_extension`
```

Comment:
[The Selenium test that is failing](https://github.com/galaxyproject/galaxy/runs/2784473904) in the `training` branch is:
```
lib/galaxy_test/selenium/test_history_dataset_state.py::HistoryDatasetStateTestCase::test_dataset_change_dbkey
```
According to https://github.com/galaxyproject/galaxy/commit/8f1e14bed8a24da6117051f13f30cd96ff62318f it is probably correct, but then this tutorial block probably needs to be updated to address it instead of the proposed ones or vice-versa.

Reply:
Another solution, which is potentially not the best idea, is to just delete the selenium test that is failing here? 

Make it seem more like a runtime error that doesn't have any existing code coverage. ¯\_(ツ)_/¯

Reply:
> Another solution, which is potentially not the best idea, is to just delete the selenium test that is failing here?
> 
> Make it seem more like a runtime error that doesn't have any existing code coverage. ¯_(ツ)_/¯

I support this idea. It's not like we're developing something for production. Our goal here is to create a small usecase, deleting them will reduce confusion and that will contribute the teaching process.

Reply:
Deleting a test is an option.. but that opens precendent: on a future update some other tests will fail due to our "installed" bugs - we would have to delete those too, etc.
If there are multiple failing tests and only some of them are relevant to a given bug, I think we can explain why there are to be expected multiple test failures - somewhere at the beginning of the tutorial (or at the beginning of that section). If the test is not available (i.e., it did not run due to some other bug we introduced), then we can just name that test.

I'm not opposed to deleting tests to accommodate the tutorial, but I think we should try to avoid this if we can.